### PR TITLE
Reduce vec macro usage

### PR DESCRIPTION
**Pull request by leichak**
Pull request for #3132.

### CODE REVIEW

File: io/src/lib.rs [hunk 1]
```diff
@@ -354,7 +354,7 @@ mod tests {
     fn read_to_limit_greater_than_total_length() {
         let s = "16-byte-string!!".to_string();
         let mut reader = Cursor::new(&s);
-        let mut buf = vec![];
+        let mut buf = Vec::with_capacity(0);
```
Comment (Kixunil) [inline io/src/lib.rs:357]: This does nothing interesting. If we can't predict the capacity we should use `Vec::new()`

File: bitcoin/embedded/src/main.rs [hunk 2]
```diff
@@ -32,14 +32,14 @@ fn main() -> ! {
     unsafe { ALLOCATOR.init(cortex_m_rt::heap_start() as usize, HEAP_SIZE) }
 
     let size = Secp256k1::preallocate_size();
-    hprintln!("secp buf size {}", size*16).unwrap();
+    hprintln!("secp buf size {}", size * 16).unwrap();
```
Comment (Kixunil) [inline bitcoin/embedded/src/main.rs:34]: It's preferable to not make unrelated changes like this.

File: bitcoin/src/network/mod.rs [hunk 3]
```diff
@@ -386,11 +386,11 @@ mod tests {
...
         use Network::*;
-        let tests = vec![
-            (Bitcoin, "bitcoin"),
-            (Testnet, "testnet"),
-            (Signet, "signet"),
-            (Regtest, "regtest"),
+        let tests = [
+            (Bitcoin, "bitcoin"), 
+            (Testnet, "testnet"), 
+            (Signet, "signet"),
```
Comment (Kixunil) [inline bitcoin/src/network/mod.rs:392]: Extra spaces.

File: hashes/Cargo.toml [hunk 4]
```diff
@@ -39,3 +39,4 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [lints.rust]
 unexpected_cfgs = { level = "deny", check-cfg = ['cfg(bench)', 'cfg(hashes_fuzz)', 'cfg(rust_v_1_64)' ] }
+
```
Comment (Kixunil) [inline hashes/Cargo.toml:42]: unrelated

**Review by Kixunil — COMMENTED** [review]
Mostly good. I'd prefer not to use `with_capacity(0)` anywhere. While I myself use `Vec::new` I think it'd be reasonable to just keep it as `vec![]`.

**Comment by tcharding** [comment]
Hi @leichak, thanks for your contribution! New devs are welcome but please read the `CONTRIBUTING.md` file's [preparing PRs'](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/CONTRIBUTING.md#preparing-prs) section. Especially please read the two blog posts linked from that section and apply what you learn to this PR. That will help us review your PRs and help get your changes merged.

Thanks again.

**Comment by leichak** [comment]
@Kixunil Wondering, so why revert to vec![] instaed of Vec::new()?, bc its shorter? 

Thanks

**Comment by apoelstra** [comment]
Please rebase on master, drop the format-only commits (though I appreciate them being in separate commits so they're easy to drop) and remove merge commits from the PR.

File: hashes/Cargo.toml [hunk 5]
```diff
@@ -38,5 +38,4 @@ all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 
 [lints.rust]
-unexpected_cfgs = { level = "deny", check-cfg = ['cfg(bench)', 'cfg(hashes_fuzz)', 'cfg(rust_v_1_64)' ] }
-
+unexpected_cfgs = { level = "deny", check-cfg = ['cfg(bench)', 'cfg(hashes_fuzz)', 'cfg(rust_v_1_64)' ] }
```
Comment (Kixunil) [inline hashes/Cargo.toml:41]: Unrelated.

**Review by Kixunil — COMMENTED** [review]
Please squash the commits

**Comment by apoelstra** [comment]
lol what is going on with this PR

**Comment by leichak** [comment]
> lol what is going on with this PR

haha, sorry, I messed up something with my git and had to fix it, but just have squashed everything as required. Also read PR guidelines for future contributions as well.

**Comment by apoelstra** [comment]
Can you fix the author and the message on the squashed commit?

File: bitcoin/src/pow.rs [hunk 6]
```diff
@@ -1930,7 +1930,7 @@ mod tests {
     #[test]
     fn work_log2() {
         // Compare work log2 to historical Bitcoin Core values found in Core logs.
-        let tests: Vec<(u128, f64)> = vec![
+        let tests: [(u128, f64); 5] = [
```
Comment (Kixunil) [inline bitcoin/src/pow.rs:1933]: I wonder if you can write `_` instead of `5` so that people won't need to change the type if they add a test.

File: hashes/Cargo.toml [hunk 7]
```diff
@@ -38,4 +38,4 @@ all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 
 [lints.rust]
-unexpected_cfgs = { level = "deny", check-cfg = ['cfg(bench)', 'cfg(hashes_fuzz)', 'cfg(rust_v_1_64)' ] }
+unexpected_cfgs = { level = "deny", check-cfg = ['cfg(bench)', 'cfg(hashes_fuzz)', 'cfg(rust_v_1_64)' ] }
```
Comment (Kixunil) [inline hashes/Cargo.toml:41]: This is still here.

File: hashes/src/ripemd160.rs [hunk 8]
```diff
@@ -481,18 +479,8 @@ mod tests {
...
-                    <&[u8; 20]>::try_from(&*test.output).expect("known length")
-                ),
-                &hash
-            );
-            assert_eq!(
-                ripemd160::Hash::from_bytes_mut(
-                    <&mut [u8; 20]>::try_from(&mut *test.output).expect("known length")
-                ),
-                &hash
-            );
```
Comment (Kixunil) [inline hashes/src/ripemd160.rs:495]: Nice! This is an actual example where `vec!` was ugly.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Apart from one unrelated change, the commit message has uninformative title and the whole message appears to be excessively long. It contains unimportant messages and also messages from previous commits. It definitely needs to be reworded to something like "Removed unneeded usages of `vec!` macro."

**Comment by tcharding** [comment]
PR title, and PR description also need editing.

File: bitcoin/src/pow.rs (hunk 6 above)
Comment (leichak) [reply bitcoin/src/pow.rs:1933]: This does not compile.

Comment (Kixunil) [reply bitcoin/src/pow.rs:1933]: OK, thanks for looking at it!

**Review by Kixunil — APPROVED** [review]
ACK 1518efc9a16dc8932821399efdc8349c73602c1f

File: bitcoin/src/pow.rs (hunk 6 above)
Comment (antonilol) [reply bitcoin/src/pow.rs:1933]: With a slice it is possible to leave out the length: `let tests: &[(u128, f64)] = &[ ...`. Now the `vec!` macro is gone here, can `#[cfg(feature = "std")]` be removed?

File: hashes/src/hash160.rs [hunk 9]
```diff
@@ -57,16 +57,16 @@ mod tests {
         #[derive(Clone)]
         #[cfg(feature = "alloc")]
         struct Test {
-            input: Vec<u8>,
-            output: Vec<u8>,
+            input: [u8; 65],
+            output: [u8; 20],
             output_str: &'static str,
         }
```
Comment (antonilol) [inline hashes/src/hash160.rs:63]: Not dependent on `alloc` anymore, `#[cfg(feature = "alloc")]` may be removed. I think there are more of these cases in this PR.

File: hashes/src/hmac.rs [hunk 10]
```diff
@@ -169,70 +169,70 @@ mod tests {
...
 
         #[derive(Clone)]
-        struct Test {
-            key: Vec<u8>,
-            input: Vec<u8>,
-            output: Vec<u8>,
+        struct Test<'a> {
+            key: &'a [u8],
+            input: &'a [u8],
+            output: &'a [u8],
```
Comment (antonilol) [inline hashes/src/hmac.rs:175]: Can `output` be a fixed size array?

File: bitcoin/src/network/mod.rs [hunk 11]
```diff
@@ -386,11 +386,11 @@ mod tests {
     #[cfg(feature = "serde")]
     fn serde_roundtrip() {
         use Network::*;
-        let tests = vec![
+        let tests = [
             (Bitcoin, "bitcoin"),
             (Testnet, "testnet"),
             (Signet, "signet"),
-            (Regtest, "regtest"),
+            (Regtest, "regtest")
```
Comment (apoelstra) [inline bitcoin/src/network/mod.rs:393]: In 1518efc9a16dc8932821399efdc8349c73602c1f

Can you put this comma back?

File: hashes/src/hash160.rs (hunk 9 above)
Comment (leichak) [reply hashes/src/hash160.rs:63]: Maybe thats the thing for another PR?

File: hashes/src/hmac.rs (hunk 10 above)
Comment (leichak) [reply hashes/src/hmac.rs:175]: As I can see could be, what about more tests in the future?

File: bitcoin/src/network/mod.rs (hunk 11 above)
Comment (leichak) [reply bitcoin/src/network/mod.rs:393]: Sure

File: bitcoin/src/pow.rs (hunk 6 above)
Comment (Kixunil) [reply bitcoin/src/pow.rs:1933]: FTR slice would make the code a bit annoying having to dereference the references.

**Review by Kixunil — APPROVED** [review]
ACK 2756b7fd7a534e7f933fbd6bad7408b5de21d9a6

**Review by apoelstra — APPROVED** [review]
ACK 2756b7fd7a534e7f933fbd6bad7408b5de21d9a6 successfully ran local tests
